Mathlib/NumberTheory/FLT/Three.lean

```diff
@@ -3,38 +3,122 @@ Copyright (c) 2024 Riccardo Brasca. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Riccardo Brasca
 -/
+import Mathlib.NumberTheory.FLT.Basic
 import Mathlib.Data.ZMod.Basic
+import Mathlib.RingTheory.Int.Basic
 
 /-!
 # Fermat Last Theorem in the case `n = 3`
 The goal of this file is to prove Fermat Last theorem in the case `n = 3`.
 
 ## Main results
-* `fermatLastTheoremThree_case_1`: the first case of Fermat Last Theorem when `n = 3`:
+* `fermatLastTheoremThree_case1`: the first case of Fermat Last Theorem when `n = 3`:
   if `a b c : ℕ` are such that `¬ 3 ∣ a * b * c`, then `a ^ 3 + b ^ 3 ≠ c ^ 3`.
 
 ## TODO
 Prove case 2.
 -/
 
+section case1
+
 open ZMod
 
 private lemma cube_of_castHom_ne_zero {n : ZMod 9} :
     castHom (show 3 ∣ 9 by norm_num) (ZMod 3) n ≠ 0 → n ^ 3 = 1 ∨ n ^ 3 = 8 := by
   fin_cases n <;> decide
 
-private lemma cube_of_not_dvd {n : ℕ} (h : ¬ 3 ∣ n) :
+private lemma cube_of_not_dvd {n : ℤ} (h : ¬ 3 ∣ n) :
     (n : ZMod 9) ^ 3 = 1 ∨ (n : ZMod 9) ^ 3 = 8 := by
   apply cube_of_castHom_ne_zero
-  rwa [map_natCast, Ne, Fin.nat_cast_eq_zero]
+  rwa [map_intCast, Ne, ZMod.int_cast_zmod_eq_zero_iff_dvd]
 
-/--If `a b c : ℕ` are such that `¬ 3 ∣ a * b * c`, then `a ^ 3 + b ^ 3 ≠ c ^ 3`. -/
-theorem fermatLastTheoremThree_case_1 {a b c : ℕ} (hdvd : ¬ 3 ∣ a * b * c) :
+/--
+  Let `a`, `b` and `c` be in `ℤ`.
+  If `¬ 3 ∣ a * b * c`, then `a ^ 3 + b ^ 3 ≠ c ^ 3`.
+-/
+theorem fermatLastTheoremThree_case1 {a b c : ℤ} (hdvd : ¬ 3 ∣ a * b * c) :
     a ^ 3 + b ^ 3 ≠ c ^ 3 := by
-  simp_rw [Nat.prime_three.dvd_mul, not_or] at hdvd
-  apply mt (congrArg (Nat.cast : ℕ → ZMod 9))
-  simp_rw [Nat.cast_add, Nat.cast_pow]
+  simp_rw [Int.prime_three.dvd_mul, not_or] at hdvd
+  apply mt (congrArg (Int.cast : ℤ → ZMod 9))
+  simp_rw [Int.cast_add, Int.cast_pow]
   rcases cube_of_not_dvd hdvd.1.1 with ha | ha <;>
   rcases cube_of_not_dvd hdvd.1.2 with hb | hb <;>
   rcases cube_of_not_dvd hdvd.2 with hc | hc <;>
   rw [ha, hb, hc] <;> decide
+
+end case1
+
+section case2
+
+private lemma three_dvd_c_of_dvd_a_of_dvd_b {a b c : ℤ} (HF : a ^ 3 + b ^ 3 + c ^ 3 = 0)
+    (h3a : 3 ∣ a) (h3b : 3 ∣ b) : 3 ∣ c := by
```

Comment:
This can be stated for all exponents `n` and all integers `3` (and might be useful more generally).
It can then also be used to golf `isCoprime_of_gcd_eq_one_of_FLT`.

Reply:
Done.